fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/impl/attachment/AttachmentSerializingImpl.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.attachment;
+
+import java.util.IdentityHashMap;
+import java.util.Map;
+
+import com.mojang.datafixers.util.Pair;
+import com.mojang.serialization.Codec;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.nbt.NbtCompound;
+import net.minecraft.nbt.NbtElement;
+import net.minecraft.nbt.NbtOps;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.attachment.v1.AttachmentTarget;
+import net.fabricmc.fabric.api.attachment.v1.AttachmentType;
+
+public class AttachmentSerializingImpl {
+	@SuppressWarnings("unchecked")
+	public static void serializeAttachmentData(NbtCompound nbt, @Nullable IdentityHashMap<AttachmentType<?>, ?> attachments) {
+		if (attachments == null) {
+			return;
+		}
+
+		var compound = new NbtCompound();
+
+		for (Map.Entry<AttachmentType<?>, ?> entry : attachments.entrySet()) {
+			AttachmentType<?> type = entry.getKey();
+
+			if (type.persistent()) {
+				Codec<Object> codec = (Codec<Object>) type.codec();
+				// non-nullity enforced by builder API
+				codec.encodeStart(NbtOps.INSTANCE, entry.getValue())
+						.result()
```

Comment:
Question: What happens with errors when (de)serialising? What does the error look like? Can we catch the error and possibly show something useful with the attachment type id?

Reply:
Well, that would only happen if:

1. Someone manually wrote some stuff in Fabric's attachment section and fucked it up
2. A codec was changed in a completely backwards-incompatible way (which is highly unlikely due to the modularity of Codecs and the stability of the base types, even more so in vanilla code).

Both seem pretty unlikely, but I'll try to test what kind of error we'd get.

Reply:
Maybe use yarn `Util::getResult`/mojmap `Util::getOrThrow` like vanilla and `FabricRecipeProvider::run`?

Reply:
Whatever the error looks like, I have changed the implementation to display it along with a warning that the attachment has been skipped altogether.